File: app/src/main/java/com/lence/penguinsvskillerwhales/life/Eat.java
package com.lence.penguinsvskillerwhales.life;

import android.graphics.Point;

import com.lence.penguinsvskillerwhales.model.Orca;
import com.lence.penguinsvskillerwhales.model.Penguin;
import com.lence.penguinsvskillerwhales.utils.Constants;
import com.lence.penguinsvskillerwhales.view.KotlinField;
import com.lence.penguinsvskillerwhales.view.UpdatePresenter;

public class Eat {
    private final KotlinField mField;
    private final Point mCurrent;
    private UpdatePresenter mUpdatePresenter;
    private Orca mOrca;

    Eat(KotlinField mField, Point mCurrent, UpdatePresenter mUpdatePresenter, Orca mOrca) {
        this.mField = mField;
        this.mCurrent = mCurrent;
        this.mUpdatePresenter = mUpdatePresenter;
        this.mOrca = mOrca;
    }

    public void execute() {
        //check of all directions

        for (int direction : Constants.directions) {
            if (findPenguin(direction)) {
                eat(KotlinField.getPointByDirection(mCurrent.x, mCurrent.y,direction));
                mOrca.eat();
                mUpdatePresenter.update(mField);
            }
        }
        mOrca.addHunger();
        new Move(mUpdatePresenter, mField,mCurrent).execute();
    }


    private void eat(Point to) {
        mField.set(to.x,to.y, mField.get(mCurrent.x, mCurrent.y));
        mField.set(mCurrent.x, mCurrent.y, null);
    }

    private boolean findPenguin(int direction) {
        Point p = KotlinField.getPointByDirection(mCurrent.x, mCurrent.y, direction);
        return mField.isInField(p.x, p.y) && (mField.get(p.x, p.y) instanceof Penguin);
    }
}

File: app/src/main/java/com/lence/penguinsvskillerwhales/life/Eat.kt
package com.lence.penguinsvskillerwhales.life

import android.graphics.Point

import com.lence.penguinsvskillerwhales.model.Orca
import com.lence.penguinsvskillerwhales.model.Penguin
import com.lence.penguinsvskillerwhales.utils.Constants
import com.lence.penguinsvskillerwhales.view.Field
import com.lence.penguinsvskillerwhales.view.UpdatePresenter

class Eat internal constructor(private val mField: Field,
                               private val mCurrent: Point,
                               private val mUpdatePresenter: UpdatePresenter,
                               private val mOrca: Orca) {

    fun execute() {
        //check of all directions

        for (direction in Constants.directions) {
            if (findPenguin(direction)) {
                eat(Field.getPointByDirection(mCurrent.x, mCurrent.y, direction))
                mOrca.eat()
                mUpdatePresenter.update(mField)
            }
        }
        mOrca.addHunger()
        Move(mUpdatePresenter, mField, mCurrent).execute()
    }


    private fun eat(to: Point) {
        mField[to.x, to.y] = mField[mCurrent.x, mCurrent.y]
        mField[mCurrent.x, mCurrent.y] = null
    }

    private fun findPenguin(direction: Int): Boolean {
        val p = Field.getPointByDirection(mCurrent.x, mCurrent.y, direction)
        return mField.isInField(p.x, p.y) && mField[p.x, p.y] is Penguin
    }
}

